### PR TITLE
Reorganise unit tests of fitting functions:: 'General' Category - Chunk 3

### LINKED ISSUE #16485: Reorganize unit tests of functions, category "General", chunk 3
This is one chunk of #16267.
Move the test cases that use the algorithm Fit to the unit test file of Fit (FitTest). Add tests to test the function itself when necessary. Tidy up the includes wherever possible, as for example in #16420.

Functions in this chunk:

- ProductFunction  :ballot_box_with_check: 
- ReflectivityMulf :no_entry_sign: 
- Resolution :ballot_box_with_check:  
- StretchExp :ballot_box_with_check: 
- TabulatedFunction :no_entry_sign: 
- ThermalNeutronBk2BkExpAlpha :no_entry_sign: 
- ThermalNeutronBk2BkExpBeta :no_entry_sign: 
- ThermalNeutronBk2BkExpConvPVoigt :no_entry_sign:

### PR DESCRIPTION
**Pull request by shahrozahmed1**
**Description of work.**

All unit tests using `Fit` algorithm from General Category have been moved to `FitTest.h`. Most of the files mentioned in the issue were not utilising the `Fit` algorithm. 

**To test:**
Check all build pass
Check all test pass locally without any issue
Check that the following files are not use the `Fit` algorithm
- ReflectivityMulf 
- TabulatedFunction
- ThermalNeutronBk2BkExpAlpha
- ThermalNeutronBk2BkExpBet
- ThermalNeutronBk2BkExpConvPVoigt

_Does not need to be in the release notes._
Fixes #16485

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 1]
```diff
@@ -538,6 +541,259 @@ class FitTest : public CxxTest::TestSuite {
...
+  void test_function_ProductFunction() {
+
+    // Mock data
+    int ndata = 30;
+    API::MatrixWorkspace_sptr ws = API::WorkspaceFactory::Instance().create(
+        "Workspace2D", 1, ndata, ndata);
+    Mantid::MantidVec &x = ws->dataX(0);
+    Mantid::MantidVec &y = ws->dataY(0);
+    Mantid::MantidVec &e = ws->dataE(0);
+    // values extracted from y(x)=2*exp(-(x/4)^0.5)
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: From the formula, y(0) = 2, but below y(0) = 0.001362. Which is correct?

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 2]
```diff
@@ -538,6 +541,259 @@ class FitTest : public CxxTest::TestSuite {
...
+    x = {0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9,
+         1, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9,
+         2, 2.1, 2.2, 2.3, 2.4, 2.5, 2.6, 2.7, 2.8, 2.9};
+    y = {0.001362, 0.00434468, 0.0127937, 0.0347769, 0.0872653, 0.202138,
+         0.432228, 0.853165,   1.55457,   2.61483,   4.06006,   5.8194,
+         7.69982,  9.40459,    10.6036,   11.0364,   10.6036,   9.40459,
+         7.69982,  5.8194,     4.06006,   2.61483,   1.55457,   0.853165,
+         0.432228, 0.202138,   0.0872653, 0.0347769, 0.0127937, 0.00434468};
+    e = {0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1,
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: This is equal to `e.assign(ndata,0.1)`

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 3]
```diff
@@ -538,6 +541,259 @@ class FitTest : public CxxTest::TestSuite {
...
+        5.433072, 5.789067, 6.089806, 6.324555, 6.484675, 6.564144, 6.559937,
+        6.472215, 6.304315, 6.062539, 5.755762, 5.394893, 4.992230, 4.560768,
+        4.113514, 3.662855, 3.220017, 2.794656, 2.394584, 2.025646, 1.691721,
+        1.394844, 1.135414, 0.912461, 0.723946, 0.567061, 0.438516, 0.334790,
+        0.252343, 0.187776, 0.137950, 0.100055, 0.071644, 0.050648, 0.035348,
+        0.024356, 0.016568, 0.011127, 0.007378, 0.004829, 0.003121, 0.001991,
+        0.001254, 0.000780, 0.000479, 0.000290, 0.000174, 0.000103, 0.000060,
+        0.000034,
+    };
+    E = {
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: This is equal to `E.assign(nY,1.0);`

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 4]
```diff
@@ -538,6 +541,259 @@ class FitTest : public CxxTest::TestSuite {
...
+    y[16] = 0.27067057;
+    y[17] = 0.25451242;
+    y[18] = 0.2397465;
+    y[19] = 0.22620756;
+
+    for (int i = 0; i <= 19; i++) {
+      // estimate errors as ten percent of the "count number"
+      const double cc = 0.1;
+      e[i] = cc * y[i];
+    }
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: Is this more clear as a `std::transform`?

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 5]
```diff
@@ -538,6 +541,259 @@ class FitTest : public CxxTest::TestSuite {
...
+    // create mock data to test against
+    std::string wsName = "StretchExpMockData";
+    int histogramNumber = 1;
+    int timechannels = 20;
+    Workspace_sptr ws = WorkspaceFactory::Instance().create(
+        "Workspace2D", histogramNumber, timechannels, timechannels);
+    Workspace2D_sptr ws2D = boost::dynamic_pointer_cast<Workspace2D>(ws);
+    // in this case, x-values are just the running index
+    for (int i = 0; i < 20; i++)
+      ws2D->dataX(0)[i] = 1.0 * i + 0.00001;
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: In general we want to obtain a reference to `dataX(0)` before entering the loop.

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 6]
```diff
@@ -538,6 +541,259 @@ class FitTest : public CxxTest::TestSuite {
...
+    y[11] = 0.38092026;
+    y[12] = 0.35384241;
+    y[13] = 0.32968143;
+    y[14] = 0.30799199;
+    y[15] = 0.28841799;
+    y[16] = 0.27067057;
+    y[17] = 0.25451242;
+    y[18] = 0.2397465;
+    y[19] = 0.22620756;
+
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: Could this be replaced with an initializer list or calculated during the test?

**Comment by shahrozahmed1** [comment]
thank you for the feedback @quantumsteve, previous commit should fix these issues. Please let me know if there is anything else.

File: Framework/CurveFitting/test/Algorithms/FitTest.h (hunk 1 above)
Comment (shahrozahmed1) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: I think this was just a copy and paste error by me

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 7]
```diff
@@ -572,6 +575,220 @@ class FitTest : public CxxTest::TestSuite {
...
+  void getStretchExpMockData(Mantid::MantidVec &y, Mantid::MantidVec &e) {
+    // values extracted from y(x)=2*exp(-(x/4)^0.5)
+    y = {2,          1.2130613,  0.98613738, 0.84124005, 0.73575888,
+         0.65384379, 0.58766531, 0.53273643, 0.48623347, 0.44626032,
+         0.41148132, 0.38092026, 0.35384241, 0.32968143, 0.30799199,
+         0.28841799, 0.27067057, 0.25451242, 0.2397465,  0.22620756};
+
+    std::transform(y.begin(), y.end(), e.begin(),
+                   std::bind2nd(std::multiplies<double>(), 0.1));
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: `std::bind2nd` has been deprecated in favor of `std::bind` and lambda expressions.

File: Framework/CurveFitting/test/Algorithms/FitTest.h [hunk 8]
```diff
@@ -572,6 +575,220 @@ class FitTest : public CxxTest::TestSuite {
...
+    double resH = 3;
+    double yErr = 0;
+    double y0 = 0;
+
+    for (int i = 0; i < N; i++) {
+      double x = X0 + i * dX;
+      double y = resH * exp(-x * x * resS);
+      double err = fabs(y - y0) / 10;
+      if (err > yErr)
+        yErr = err;
```
Comment (quantumsteve) [inline Framework/CurveFitting/test/Algorithms/FitTest.h]: `std::max`?

**Comment by quantumsteve** [comment]
Thank you for for taking the extra time to not only reorganize, but also clean up these tests. Once the deprecated `bind2nd` function is replaced, I think it is ready to merge.

**Comment by shahrozahmed1** [comment]
@quantumsteve this show now be ready, thanks

**Comment by quantumsteve** [comment]
LGTM :shipit:

**Pull request merged**